metaflow/plugins/kfp/kfp_s3_sensor.py

```diff
@@ -0,0 +1,62 @@
+"""
+This function is called within the s3_sensor_op running container.
+(1) It decodes the path_formatter function's code and runs it to obtain a final S3 path
+(2) It splits the formatted path into an S3 bucket and key 
+(3) It polls for an object with the specified bucket and key until timeout
+"""
+import boto3
+
+
+def wait_for_s3_path(
+    path: str,
+    timeout_seconds: int,
+    polling_interval_seconds: int,
+    path_formatter_code_encoded: str,
+    flow_parameters_json: str,
+    os_vars: bool,
+) -> str:
+    import boto3
+    import botocore
+    import base64
+    import json
+    import marshal
+    import time
+    from typing import Tuple
+    from urllib.parse import urlparse
+    import os
+
+    flow_parameters = json.loads(flow_parameters_json)
+    path_formatter_code = marshal.loads(base64.b64decode(path_formatter_code_encoded))
+
+    def path_formatter_template(key: str, flow_parameters: dict) -> str:
+        pass
+
+    path_formatter_template.__code__ = path_formatter_code
+    path = path_formatter_template(path, flow_parameters)
+    if os_vars:
+        # expand OS env variables
+        path = os.path.expandvars(path)
+    # default variable substitution
+    path = path.format(**flow_parameters)
+    parsed_path = urlparse(path)
+    bucket, key = parsed_path.netloc, parsed_path.path.lstrip("/")
+
+    s3 = boto3.client("s3")
+    start_time = time.time()
+    while True:
+        current_time = time.time()
+        elapsed_time = current_time - start_time
+        if timeout_seconds is not -1 and elapsed_time > timeout_seconds:
```

Comment:
Please consider remove the -1 logic for indefinite waiting: we will not want our code to indefinitely wait on some s3 path if the path never exists (maybe a typo in s3). 

We can even considering setting an upper limit for timeout, and warn customer they are setting too long a timeout when they exceeds the limit. 48 hour (to be discussed) may be a good limit in my opinion.

Reply:
@cloudw actually the default is a positive number, so the customer won't fall into this trap unless the explicitly set it to -1. I want to give them an option to wait infinity if they really want to.

Reply:
But the idea of a warning sounds good. I can definitely do that.

Reply:
I understand the tendency to provide as much features as possible, but I maintain my opinion that allowing indefinite wait can be either dangerous or resource consuming. I also think that undocumented features should not be maintained 🙂 .

If you do think this option should be available, please add it to decorator doc string so that this feature is known.

Reply:
Actually @cloudw thinking more about it, I think you're right in not allowing waiting for more than 1 or 2 days :) Better to prevent the worst before it happens, so I'm happy to make a check to raise an exception if the customer ask for a timeout more than 1 day.

Can we get a second opinion on this @talebzeghmi ?